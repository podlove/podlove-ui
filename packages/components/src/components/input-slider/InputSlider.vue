<script lang="ts" setup>
import { computed } from 'vue';
import { path } from 'ramda';
import { round } from '@podlove/utils/math';

const pinRange = 0.01;

const props = defineProps({
  min: {
    type: Number,
    default: 0
  },
  max: {
    type: Number,
    default: 100
  },
  step: {
    type: Number,
    default: 0.1
  },
  value: {
    type: Number,
    default: 0
  },
  pins: {
    type: Array<{value: number, label: string}>,
    default: () => []
  }
});

const emit = defineEmits(['sliderInput', 'sliderChange', 'sliderDblclick'])

const relativePosition = (current: any = 0, minimum: any = 0, maximum: any = 0) =>
  ((parseFloat(current) - parseFloat(minimum)) * 100) /
    (parseFloat(maximum) - parseFloat(minimum)) +
  '%';

const thumbLeft = computed(() => relativePosition(props.value, props.min, props.max));

const calcValue = (event: Event) => {
      const value = path(['target','value'], event) as number;

      return (
        props.pins
          .map(({value}) => value)
          .find((pin) => pin + pinRange >= value && pin - pinRange <= value) || value
      );
    };

const handleInput = (event: Event) => {
  emit('sliderInput', calcValue(event));
};

const handleChange = (event: Event) => {
  emit('sliderChange', calcValue(event));
};

const handleDblclick = (event: Event) => {
  emit('sliderDblclick', path(['target','value'], event));
};
</script>

<template>
  <div class="podlove-component--input-slider w-full relative h-[30px]">
    <input
      type="range"
      :min="min"
      :max="max"
      :value="value"
      :step="step"
      @input="handleInput"
      @change="handleChange"
      @dblclick="handleDblclick"
    />
    <span
      class="track block w-full absolute left-0 pointer-events-none top-[50%] h-[2px] rounded-sm"
    />
    <span
      v-for="(pin, index) in pins"
      :key="index"
      class="pin block absolute font-medium text-xs text-center -top-[10px] w-[30px] -ml-[15px]"
      :style="{ left: `${round(pin.value * 100)}%` }"
    >
      {{ pin.label }}
    </span>
    <span
      class="
        thumb
        absolute
        border border-solid
        pointer-events-none
        rounded-lg
        top-[calc(50%-4px)]
        h-[10px]
        w-[10px]
        -ml-[7.5px]
      "
    />
    <slot />
  </div>
</template>

<style lang="scss" scoped>
@import '../../styles/range.scss';

.podlove-component--input-slider .thumb {
  left: v-bind('thumbLeft');
  background-color: var(
    --podlove-component--input-slider--thumb-color,
    var(--podlove-components-background)
  );
  border-color: var(
    --podlove-component--input-slider--border-color,
    var(--podlove-components-text)
  );
}

.podlove-component--input-slider .track {
  background-color: var(
    --podlove-component--input-slider--progress-color,
    var(--podlove-components-text)
  );
}

.podlove-component--input-slider {
  @include range($progress-height, $thumb-width-desktop, $thumb-width-desktop-hover);
}
</style>
