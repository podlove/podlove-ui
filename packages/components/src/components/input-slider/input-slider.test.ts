import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import InputSlider from './InputSlider.vue';

describe('InputSlider', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(InputSlider);

      expect(wrapper.vm).toBeTruthy();
    });
  });

  describe('props', () => {
    test('should render without props', async () => {
      const wrapper = await mount(InputSlider);
      expect(wrapper.element).toMatchSnapshot();
    });

    const props = [
      {
        prop: 'step',
        value: 1
      },
      {
        prop: 'value',
        value: 3
      },
      {
        prop: 'pins',
        value: [
          {
            value: 1,
            label: 'pin-label-1'
          },
          {
            value: 4,
            label: 'pin-label-2'
          }
        ]
      },
      {
        prop: 'background',
        value: '#fff'
      },
      {
        prop: 'borderColor',
        value: '#fff'
      },
      {
        prop: 'progressColor',
        value: '#fff'
      }
    ];

    props.forEach(({ prop, value }) => {
      test(`should render type '${prop}'`, async () => {
        const wrapper = await mount(InputSlider, {
          props: {
            min: 1,
            max: 20,
            [prop]: value
          }
        });

        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('events', () => {
    test('should dispatch input event on input', async () => {
      const wrapper = await mount(InputSlider, {
        props: {
          min: 1,
          max: 20,
          value: 1
        }
      });

      wrapper.find('input').trigger('input');
      expect(wrapper.emitted().sliderInput).toEqual([[1]]);
    });

    test('should dispatch change event on change', async () => {
      const wrapper = await mount(InputSlider, {
        props: {
          min: 1,
          max: 20,
          value: 1
        }
      });

      wrapper.find('input').trigger('change');
      expect(wrapper.emitted().sliderChange).toEqual([[ 1 ]]);
    });

    test('should dispatch dblclick event on dblclick', async () => {
      const wrapper = await mount(InputSlider, {
        props: {
          min: 1,
          max: 20,
          value: 1
        }
      });

      wrapper.find('input').trigger('dblclick');
      expect(wrapper.emitted().sliderDblclick).toEqual([[1]]);
    });
  });
});
