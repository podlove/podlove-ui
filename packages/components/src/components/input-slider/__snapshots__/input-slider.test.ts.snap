// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputSlider > props > should render type 'background' 1`] = `
<div
  background="#fff"
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
>
  <input
    data-v-aad1c04c=""
    max="20"
    min="1"
    step="0.1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;

exports[`InputSlider > props > should render type 'borderColor' 1`] = `
<div
  bordercolor="#fff"
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
>
  <input
    data-v-aad1c04c=""
    max="20"
    min="1"
    step="0.1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;

exports[`InputSlider > props > should render type 'pins' 1`] = `
<div
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
>
  <input
    data-v-aad1c04c=""
    max="20"
    min="1"
    step="0.1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  <span
    class="pin block absolute font-medium text-xs text-center -top-[10px] w-[30px] -ml-[15px]"
    data-v-aad1c04c=""
    style="left: 100%;"
  >
    pin-label-1
  </span>
  <span
    class="pin block absolute font-medium text-xs text-center -top-[10px] w-[30px] -ml-[15px]"
    data-v-aad1c04c=""
    style="left: 400%;"
  >
    pin-label-2
  </span>
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;

exports[`InputSlider > props > should render type 'progressColor' 1`] = `
<div
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
  progresscolor="#fff"
>
  <input
    data-v-aad1c04c=""
    max="20"
    min="1"
    step="0.1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;

exports[`InputSlider > props > should render type 'step' 1`] = `
<div
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
>
  <input
    data-v-aad1c04c=""
    max="20"
    min="1"
    step="1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;

exports[`InputSlider > props > should render type 'value' 1`] = `
<div
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
>
  <input
    data-v-aad1c04c=""
    max="20"
    min="1"
    step="0.1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;

exports[`InputSlider > props > should render without props 1`] = `
<div
  class="podlove-component--input-slider w-full relative h-[30px]"
  data-v-aad1c04c=""
>
  <input
    data-v-aad1c04c=""
    max="100"
    min="0"
    step="0.1"
    type="range"
  />
  <span
    class="track block w-full absolute left-0 pointer-events-none top-1/2 h-[2px] rounded-sm"
    data-v-aad1c04c=""
  />
  
  
  <span
    class="thumb absolute border border-solid pointer-events-none rounded-lg top-[calc(50%-4px)] h-[10px] w-[10px] -ml-[7.5px]"
    data-v-aad1c04c=""
  />
  
  
</div>
`;
